lib/scheduler.ml

```diff
@@ -426,26 +424,60 @@ module Make (Streamd : StreamDescriptor) = struct
            * inactive after the call to `write` above. *)
           let subtree_is_active =
             Streamd.requires_output descriptor
-            || not (PriorityQueue.is_empty p.children)
+            || not (PriorityQueue.is_empty children)
           in
           written, subtree_is_active
-        else (
-          match PriorityQueue.pop p.children with
-          | Some ((id, (Stream i as i_node)), children') ->
-            p.t_last <- i.t;
-            let written, subtree_is_active = schedule i_node in
+        else traverse p_node
+    and traverse : type a. a node -> int * bool =
+     fun p_node ->
+      let rec loop remaining_children =
+        match PriorityQueue.pop remaining_children with
+        | Some ((id, (Stream i as i_node)), remaining_children') ->
+          let written, subtree_is_active = schedule i_node in
+
+          if not subtree_is_active
+          then (
+            implicitly_close_idle_stream i.descriptor max_seen_ids;
+            (* XXX(anmonteiro): we may not want to remove from the tree right
+             * away. *)
+            remove_child p_node id);
+
+          if written > 0
+             (* We need this check because the queue contains both streams that
+              * want to send data and streams that don't want to send data.
+              * Without this check, we might end up in a situation where a stream
+              * won't be flushed until a few write operations later. *)
+          then (
+            update_t_last p_node i.t;
+            update_t i_node written;
             if subtree_is_active
```

Comment:
this case https://github.com/anmonteiro/ocaml-h2/pull/199/files#r1093672918 is fixed here: